### PR TITLE
BREAKING: CLI, react-native link: Add "demo" to ignore pattern

### PR DESCRIPTION
**Pull request by nikosch86**
We are using react-native-xmpp which references Robbie Hansons `XMPPFramework`, there has not been a problem before RN 37 because the `findProject` function was not working as expected.

With commit `157a03dd8dc8c57e3e9f6b4116f10e7152cf23ee` by @robclouth  this function also finds the demo projects in XMPPFramework

For example this one:
https://github.com/robbiehanson/XMPPFramework/tree/db543e1b484687768024ce3e78e9f287fbee4aa6/Xcode/ServerlessDemo

This will make you unable to build your project

Even though i highly discourage the way this findProject function scrapes everything and ignores based on regex, it seems like for cases like this, where dependencies of a dependency include demo projects whose names we can not control, the sensible fix would be to ignore projects containing  'demo'

### CODE REVIEW

File: local-cli/core/ios/findProject.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ const GLOB_PATTERN = '**/*.xcodeproj';
 /**
  * Regexp matching all test projects
```
Comment (mkonicek) [inline local-cli/core/ios/findProject.js]: Can you please add a comment saying what a test project is?

**Comment by mkonicek** [comment]
Checking for few random names such as "test|example|sample|demo" seems fragile but since we already check for "example" and "sample" I think checking for "demo" should work too.

Adding BREAKING because this now ignores all "demo" projects - hopefully no one relied on "demo" projects being included, but just in case.

**Comment by mkonicek** [comment]
Btw where is this ignore pattern used? Is it when running 'react-native link'?

File: local-cli/core/ios/findProject.js (hunk 1 above)
Comment (nikosch86) [inline local-cli/core/ios/findProject.js]: Since i am not the original author of this code i can not be certain what the intention was

**Comment by nikosch86** [comment]
@mkonicek exactly

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

**Comment by mkonicek** [comment]
Retrying.

**Comment by hramos** [comment]
This PR is not passing Jest tests. You can reproduce by running `npm run test`:

```
Summary of all failing tests
 FAIL  local-cli/core/__tests__/ios/findProject.spec.js
  ● ios::findProject › ignores Pods inside `ios` folder

    expect(received).toBe(expected)
    
    Expected value to be (using ===):
      "ios/DemoApp/demoProject.xcodeproj"
    Received:
      null
    
    Difference:
    
      Comparing two different types of values. Expected string but received null.
      
      at Object.it (local-cli/core/__tests__/ios/findProject.spec.js:52:29)
      at tryCallTwo (node_modules/promise/lib/core.js:45:5)
      at doResolve (node_modules/promise/lib/core.js:200:13)
      at new Promise (node_modules/promise/lib/core.js:66:3)
      at tryCallOne (node_modules/promise/lib/core.js:37:12)
      at node_modules/promise/lib/core.js:123:15
      at flush (node_modules/asap/raw.js:50:29)
      at _combinedTickCallback (internal/process/next_tick.js:95:7)
      at process._tickCallback (internal/process/next_tick.js:161:9)

  ● ios::findProject › ignores xcodeproj from example folders

    TypeError: Cannot read property 'toLowerCase' of null
      
      at Object.it (local-cli/core/__tests__/ios/findProject.spec.js:64:27)
      at tryCallTwo (node_modules/promise/lib/core.js:45:5)
      at doResolve (node_modules/promise/lib/core.js:200:13)
      at new Promise (node_modules/promise/lib/core.js:66:3)
      at tryCallOne (node_modules/promise/lib/core.js:37:12)
      at node_modules/promise/lib/core.js:123:15
      at flush (node_modules/asap/raw.js:50:29)
      at _combinedTickCallback (internal/process/next_tick.js:95:7)
      at process._tickCallback (internal/process/next_tick.js:161:9)

  ● ios::findProject › ignores xcodeproj from sample folders

    TypeError: Cannot read property 'toLowerCase' of null
      
      at Object.it (local-cli/core/__tests__/ios/findProject.spec.js:76:27)
      at tryCallTwo (node_modules/promise/lib/core.js:45:5)
      at doResolve (node_modules/promise/lib/core.js:200:13)
      at new Promise (node_modules/promise/lib/core.js:66:3)
      at tryCallOne (node_modules/promise/lib/core.js:37:12)
      at node_modules/promise/lib/core.js:123:15
      at flush (node_modules/asap/raw.js:50:29)
      at _combinedTickCallback (internal/process/next_tick.js:95:7)
      at process._tickCallback (internal/process/next_tick.js:161:9)

  ● ios::findProject › ignores xcodeproj from test folders at any level

    TypeError: Cannot read property 'toLowerCase' of null
      
      at Object.it (local-cli/core/__tests__/ios/findProject.spec.js:90:27)
      at tryCallTwo (node_modules/promise/lib/core.js:45:5)
      at doResolve (node_modules/promise/lib/core.js:200:13)
      at new Promise (node_modules/promise/lib/core.js:66:3)
      at tryCallOne (node_modules/promise/lib/core.js:37:12)
      at node_modules/promise/lib/core.js:123:15
      at flush (node_modules/asap/raw.js:50:29)
      at _combinedTickCallback (internal/process/next_tick.js:95:7)
      at process._tickCallback (internal/process/next_tick.js:161:9)


Test Suites: 1 failed, 2 skipped, 84 passed, 85 of 87 total
Tests:       4 failed, 4 skipped, 465 passed, 473 total
Snapshots:   35 passed, 35 total
Time:        9.021s, estimated 16s
Ran all test suites.
```

**Comment by nikosch86** [comment]
@hramos i changed the name of the mocked app so it does not match the regex anymore, the tests are successful now

**Pull request closed without merging**